src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -28,28 +28,9 @@ internal static partial class FileSystem
             UnixFileMode.OtherWrite |
             UnixFileMode.OtherExecute;
 
-        public static void CopyFile(string sourceFullPath, string destFullPath, bool overwrite)
-        {
-            long fileLength;
-            UnixFileMode filePermissions;
-            using SafeFileHandle src = SafeFileHandle.OpenReadOnly(sourceFullPath, FileOptions.None, out fileLength, out filePermissions);
-            using SafeFileHandle dst = SafeFileHandle.Open(destFullPath, overwrite ? FileMode.Create : FileMode.CreateNew,
-                                            FileAccess.ReadWrite, FileShare.None, FileOptions.None, preallocationSize: 0, filePermissions,
-                                            CreateOpenException);
-
-            Interop.CheckIo(Interop.Sys.CopyFile(src, dst, fileLength));
-
-            static Exception? CreateOpenException(Interop.ErrorInfo error, Interop.Sys.OpenFlags flags, string path)
-            {
-                // If the destination path points to a directory, we throw to match Windows behaviour.
-                if (error.Error == Interop.Error.EEXIST && DirectoryExists(path))
-                {
-                    return new IOException(SR.Format(SR.Arg_FileIsDirectory_Name, path));
-                }
-
-                return null; // Let SafeFileHandle create the exception for this error.
-            }
-        }
+        // CopyFile is defined in either FileSystem.CopyFile.OSX.cs or FileSystem.CopyFile.OtherUnix.cs
+        // The implementations on OSX-like Operating Systems attempts to clone the file first.
+        public static partial void CopyFile(string sourceFullPath, string destFullPath, bool overwrite);
```

Comment:
To avoid the duplication, this could be:
```diff
$ git diff
diff --git a/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs b/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
index f28cd70f08f..1ac23fdc8ed 100644
--- a/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
@@ -33,6 +33,12 @@ public static void CopyFile(string sourceFullPath, string destFullPath, bool ove
             long fileLength;
             UnixFileMode filePermissions;
             using SafeFileHandle src = SafeFileHandle.OpenReadOnly(sourceFullPath, FileOptions.None, out fileLength, out filePermissions);
+
+            if (TryOSXCloneFile(sourceFullPath, destFullPath, overwrite))
+            {
+                return;
+            }
+
             using SafeFileHandle dst = SafeFileHandle.Open(destFullPath, overwrite ? FileMode.Create : FileMode.CreateNew,
                                             FileAccess.ReadWrite, FileShare.None, FileOptions.None, preallocationSize: 0, filePermissions,
                                             CreateOpenException);

```

Reply:
I like this idea, so will be doing it, but a few Qs first.
Could this be done without requiring `TryOSXCloneFile` to exist on other unix platforms? I don't think it needing to exist elsewhere would be ideal. Maybe is there a `#if` for OSX-like plaforms (not just macOS)?

Reply:
Or maybe it could just be named `TryCloneFile`. I will implement it like this for now.

Reply:
> Could this be done without requiring TryOSXCloneFile to exist on other unix platforms?

In the `.OtherUnix.c` file you add an implementation that just returns `false`.

Reply:
My issue was more having a function named `OSX` that just returns false exist on a linux build for no reason. Hence `TryCloneFile` if you're okay with that

Reply:
I have implemented this :)

Reply:
I don't feel strongly about it. Adding OSX in the name gives an idea to the reader the function does something OSX specific without having to look deeper into the code.